### PR TITLE
sampling: add Top-nσ sampler to `llama-server`

### PR DESCRIPTION
**Pull request by CasualAutopsy**

### CODE REVIEW

File: common/sampling.cpp [hunk 1]
```diff
@@ -190,8 +190,9 @@ struct common_sampler * common_sampler_init(const struct llama_model * model, co
     if (params.mirostat == 0) {
         if (params.top_n_sigma >= 0) {
             llama_sampler_chain_add(result->chain, llama_sampler_init_top_k        (params.top_k));
-            llama_sampler_chain_add(result->chain, llama_sampler_init_temp         (params.temp));
+            llama_sampler_chain_add(result->chain, llama_sampler_init_temp_ext     (params.temp, params.dynatemp_range, params.dynatemp_exponent));
             llama_sampler_chain_add(result->chain, llama_sampler_init_top_n_sigma  (params.top_n_sigma));
+            llama_sampler_chain_add(result->chain, llama_sampler_init_xtc          (params.xtc_probability, params.xtc_threshold, params.min_keep, params.seed));
         } else {
```
Comment (ggerganov) [inline common/sampling.cpp:196]: This special-case should probably be removed and configure the sampler chain as usual. @VJHack What was the reason to do it this way?

Comment (MaggotHATE) [reply common/sampling.cpp:196]: There was a [discussion](https://github.com/ggml-org/llama.cpp/pull/11223#discussion_r1921086930) about `top_n_sigma` being added into the main chain, but it was probably lost.

This sampler is supposed to run after `temp` (`temp_ext` works too), but the current main chain has temperature after all other truncating samplers by default. If `top_n_sigma` is put in that order, it will suffer from previous samplers that may be used after it, but not before. Ideally, it's `top-k->temp->top_nsigma->*everything else*`, but that would require changing current default order of samplers.

Comment (DocShotgun) [reply common/sampling.cpp:196]: Seems like based on comments from the original author of `top_n_sigma` in the `aphrodite-engine` PR thread (https://github.com/aphrodite-engine/aphrodite-engine/pull/825), it doesn't matter whether `top_n_sigma` is used before or after `temperature`, but it will result in errors if it is used after other "alphabet" truncation samplers, with the exception of `top_k`.

Comment (VJHack) [reply common/sampling.cpp:196]: > This special-case should probably be removed and configure the sampler chain as usual. @VJHack What was the reason to do it this way?

I made this sampler an exception because of two reasons, one of which is[ stated in the PR](https://github.com/ggml-org/llama.cpp/pull/11223#:~:text=I%20added%20it%20as%20a%20stand%2Dalone%20sampler%20instead%20of%20chaining%20it%20with%20the%20common%20samplers):
1. As the authors of the paper mention, the manipulation for `top_n_sigma` is done on the[ logits pre-softmax](https://arxiv.org/html/2411.07641v1#:~:text=pre%2Dsoftmax%20logit%20distributions%2C). Some of the other samplers that would come before`top_n_sigma` (like `top_p` for example) are taking the softmax before truncating the number of tokens we can sample from. I was concerned that this would degrade the quality of this sampler.
2. `top_n_sigma` must be used after `temperature` is applied. In the paper they repeatedly mention  the temperature invariance property being one of the key characteristics of this sampler. @DocShotgun, I understand that this can be a bit confusing but temperature invariance means that the set of candidate tokens remains constant regardless of the temperature value used. According to the algorithm, temperature must be applied before `top_n_sigma`. 

I made `top_n_sigma` a special-case (stand alone) sampler because it operates on logits pre-softmax and must come after temperature is applied.

Comment (DocShotgun) [reply common/sampling.cpp:196]: Where do penalties fit into this order? IIRC most backends apply penalties as the very first sampler.

Comment (MaggotHATE) [reply common/sampling.cpp:196]: > most backends apply penalties as the very first sampler.

Not sure how significant it is for other backends, but applying penalties to the whole vocabulary affects performance in llama.cpp, so it is subject to change in the future, some day.

As for `top_n_sigma` and repetitions, I [tested](https://github.com/ggml-org/llama.cpp/pull/11223#discussion_r1921038400) is a bit with the same order as in the main chain, but theoretically it should not be needed. This aspect needs more testing, and I'm currently preoccupied by figuring out if Mistral Small 3 is good or not.

Comment (DocShotgun) [reply common/sampling.cpp:196]: > As for `top_n_sigma` and repetitions, I [tested](https://github.com/ggml-org/llama.cpp/pull/11223#discussion_r1921038400) it a bit with the same order as in the main chain, but theoretically it should not be needed. This aspect needs more testing, and I'm currently preoccupied by figuring out if Mistral Small 3 is good or not.

Hmm so at least it should function properly with penalties? It doesn't look like penalties are allowed to be used with `top_n_sigma` in this current implementation.

Comment (MaggotHATE) [reply common/sampling.cpp:196]: > It doesn't look like penalties are allowed to be used with `top_n_sigma` in this current implementation

Yes, either `repetition penalty` and `DRY` should be added to the new chain (after `top_k`), or `top_n_sigma` should be put into the main chain.

I'm not sure if changing the default chain order to `logits->top_k->penalties->temp->*everything else*` is a good idea now, so maybe it is worth putting this idea aside in a separate PR.

Comment (ggerganov) [reply common/sampling.cpp:196]: > Yes, either repetition penalty and DRY should be added to the new chain (after top_k), or top_n_sigma should be put into the main chain.

Let's not add a new chain. We should either put the top-sigma sampler in the existing chain or let the user explicitly add the sampler. Remember that this is just a default configuration, the sampler is noop by default and the user can re-configure the order of the samplers. If the sampler has to be applied before softmax, the user is responsible to construct the sampling chain to make sense.

**Comment by 0xDEADFED5** [comment]
what's the status of this?

**Comment by CISC** [comment]
Superceded by #13264
